--- a/drivers/mtd/mtdpart.c
+++ b/drivers/mtd/mtdpart.c
@@ -51,6 +51,8 @@
 	struct list_head list;
 };
 
+static bool rootfs_split = 1;
+
 static void mtd_partition_split(struct mtd_info *master, struct mtd_part *part);
 
 /*
@@ -836,7 +838,8 @@
 		return;
 
 	if (!strcmp(part->mtd.name, "rootfs")) {
-		run_parsers_by_type(part, MTD_PARSER_TYPE_ROOTFS);
+		if (rootfs_split)
+			run_parsers_by_type(part, MTD_PARSER_TYPE_ROOTFS);
 
 		rootfs_found = 1;
 	}
@@ -1123,3 +1126,12 @@
 	return PART(mtd)->master->size;
 }
 EXPORT_SYMBOL_GPL(mtd_get_device_size);
+
+static int __init no_rootfs_split(char *str)
+{
+	rootfs_split = 0;
+
+	return 0;
+}
+
+early_param("norootfssplit", no_rootfs_split);
