--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -677,17 +677,29 @@
 	{ "at45db081d", INFO(0x1f2500, 0, 64 * 1024, 16, SECT_4K) },
 
 	/* EON -- en25xxx */
+	{ "en25f80",    INFO(0x1c3114, 0, 64 * 1024,   16, SECT_4K) },
+	{ "en25q80a",   INFO(0x1c3014, 0, 64 * 1024,   16, SECT_4K) },
+	{ "en25q16",    INFO(0x1c3015, 0, 64 * 1024,   32, SECT_4K) },
+	{ "en25qh16",   INFO(0x1c7015, 0, 64 * 1024,   32, SECT_4K) },
 	{ "en25f32",    INFO(0x1c3116, 0, 64 * 1024,   64, SECT_4K) },
 	{ "en25p32",    INFO(0x1c2016, 0, 64 * 1024,   64, 0) },
-	{ "en25q32b",   INFO(0x1c3016, 0, 64 * 1024,   64, 0) },
+	{ "en25q32b",   INFO(0x1c3016, 0, 64 * 1024,   64, SECT_4K) },
+	{ "en25qh32",   INFO(0x1c7016, 0, 64 * 1024,   64, SECT_4K) },
 	{ "en25p64",    INFO(0x1c2017, 0, 64 * 1024,  128, 0) },
 	{ "en25q64",    INFO(0x1c3017, 0, 64 * 1024,  128, SECT_4K) },
-	{ "en25qh128",  INFO(0x1c7018, 0, 64 * 1024,  256, 0) },
-	{ "en25qh256",  INFO(0x1c7019, 0, 64 * 1024,  512, 0) },
+	{ "en25qh64",   INFO(0x1c7017, 0, 64 * 1024,  128, SECT_4K) },
+	{ "en25q128",   INFO(0x1c3018, 0, 64 * 1024,  256, SECT_4K) },
+	{ "en25qh128",  INFO(0x1c7018, 0, 64 * 1024,  256, SECT_4K) },
+	{ "en25qh256",  INFO(0x1c7019, 0, 64 * 1024,  512, SECT_4K) },
 	{ "en25s64",	INFO(0x1c3817, 0, 64 * 1024,  128, SECT_4K) },
 
 	/* ESMT */
-	{ "f25l32pa", INFO(0x8c2016, 0, 64 * 1024, 64, SECT_4K) },
+	{ "f25l08qa",  INFO(0x8c4014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "f25l16pa",  INFO(0x8c2115, 0, 64 * 1024,  32, SECT_4K) },
+	{ "f25l32pa",  INFO(0x8c2016, 0, 64 * 1024,  64, SECT_4K) },
+	{ "f25l32qa",  INFO(0x8c4116, 0, 64 * 1024,  64, SECT_4K) },
+	{ "f25l64qa",  INFO(0x8c4117, 0, 64 * 1024, 128, SECT_4K) },
+	{ "f25l128qa", INFO(0x8c4118, 0, 64 * 1024, 256, SECT_4K) },
 
 	/* Everspin */
 	{ "mr25h256", CAT25_INFO( 32 * 1024, 1, 256, 2, SPI_NOR_NO_ERASE | SPI_NOR_NO_FR) },
@@ -697,17 +709,32 @@
 	{ "mb85rs1mt", INFO(0x047f27, 0, 128 * 1024, 1, SPI_NOR_NO_ERASE) },
 
 	/* GigaDevice */
-	{ "gd25q32", INFO(0xc84016, 0, 64 * 1024,  64, SECT_4K) },
-	{ "gd25q64", INFO(0xc84017, 0, 64 * 1024, 128, SECT_4K) },
-	{ "gd25q128", INFO(0xc84018, 0, 64 * 1024, 256, SECT_4K) },
+	{ "gd25q80",  INFO(0xc84014, 0, 64 * 1024,   16, SECT_4K) },
+	{ "gd25q16",  INFO(0xc84015, 0, 64 * 1024,   32, SECT_4K) },
+	{ "gd25q32",  INFO(0xc84016, 0, 64 * 1024,   64, SECT_4K) },
+	{ "gd25q64",  INFO(0xc84017, 0, 64 * 1024,  128, SECT_4K) },
+	{ "gd25q128", INFO(0xc84018, 0, 64 * 1024,  256, SECT_4K) },
+	{ "gd25q256", INFO(0xc84019, 0, 64 * 1024,  512, SECT_4K) },
+	{ "gd25q512", INFO(0xc84020, 0, 64 * 1024, 1024, SECT_4K) },
 
 	/* Intel/Numonyx -- xxxs33b */
 	{ "160s33b",  INFO(0x898911, 0, 64 * 1024,  32, 0) },
 	{ "320s33b",  INFO(0x898912, 0, 64 * 1024,  64, 0) },
 	{ "640s33b",  INFO(0x898913, 0, 64 * 1024, 128, 0) },
+ 
+	/* Intel/Numonyx -- xxxs33t */
+	{ "160s33t",  INFO(0x898915, 0, 64 * 1024,  32, 0) },
+	{ "320s33t",  INFO(0x898916, 0, 64 * 1024,  64, 0) },
+	{ "640s33t",  INFO(0x898917, 0, 64 * 1024, 128, 0) },
 
 	/* ISSI */
 	{ "is25cd512", INFO(0x7f9d20, 0, 32 * 1024,   2, SECT_4K) },
+	{ "is25lp080", INFO(0x9d6014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "is25lp016", INFO(0x9d6015, 0, 64 * 1024,  32, SECT_4K) },
+	{ "is25lp032", INFO(0x9d6016, 0, 64 * 1024,  64, SECT_4K) },
+	{ "is25lp064", INFO(0x9d6017, 0, 64 * 1024, 128, SECT_4K) },
+	{ "is25lp128", INFO(0x9d6018, 0, 64 * 1024, 256, SECT_4K) },
+	{ "is25lp256", INFO(0x9d6019, 0, 64 * 1024, 512, SECT_4K) },
 
 	/* Macronix */
 	{ "mx25l512e",   INFO(0xc22010, 0, 64 * 1024,   1, SECT_4K) },
@@ -762,10 +789,6 @@
 	{ "s25sl016a",  INFO(0x010214,      0,  64 * 1024,  32, 0) },
 	{ "s25sl032a",  INFO(0x010215,      0,  64 * 1024,  64, 0) },
 	{ "s25sl064a",  INFO(0x010216,      0,  64 * 1024, 128, 0) },
-	{ "s25fl004k",  INFO(0xef4013,      0,  64 * 1024,   8, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
-	{ "s25fl008k",  INFO(0xef4014,      0,  64 * 1024,  16, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
-	{ "s25fl016k",  INFO(0xef4015,      0,  64 * 1024,  32, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
-	{ "s25fl064k",  INFO(0xef4017,      0,  64 * 1024, 128, SECT_4K) },
 	{ "s25fl132k",  INFO(0x014016,      0,  64 * 1024,  64, SECT_4K) },
 	{ "s25fl164k",  INFO(0x014017,      0,  64 * 1024, 128, SECT_4K) },
 	{ "s25fl204k",  INFO(0x014013,      0,  64 * 1024,   8, SECT_4K | SPI_NOR_DUAL_READ) },
@@ -836,6 +859,7 @@
 	{ "w25q128fw", INFO(0xef6018, 0, 64 * 1024, 256, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
 	{ "w25q80", INFO(0xef5014, 0, 64 * 1024,  16, SECT_4K) },
 	{ "w25q80bl", INFO(0xef4014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "w25q16", INFO(0xef4015, 0, 64 * 1024,  32, SECT_4K) },
 	{ "w25q128", INFO(0xef4018, 0, 64 * 1024, 256, SECT_4K) },
 	{ "w25q256", INFO(0xef4019, 0, 64 * 1024, 512, SECT_4K) },
 
