--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -179,6 +179,7 @@
 	ATH79_MACH_TL_WR842N_V2,	/* TP-LINK TL-WR842N/ND v2 */
 	ATH79_MACH_TL_WR941ND,		/* TP-LINK TL-WR941ND */
 	ATH79_MACH_TL_WR941ND_V5,	/* TP-LINK TL-WR941ND v5 */
+	ATH79_MACH_TL_WR941N_V6,	/* TP-LINK TL-WR941N v6 */
 	ATH79_MACH_TL_WR941N_V7,	/* TP-LINK TL-WR941N v7 */
 	ATH79_MACH_TUBE2H,		/* Alfa Network Tube2H */
 	ATH79_MACH_UBNT_AIRGW,		/* Ubiquiti AirGateway */
