--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -1421,6 +1421,16 @@
 	select ATH79_DEV_USB
 	select ATH79_NVRAM
 
+config ATH79_MACH_TL_WR882N_V1
+	bool "TP-LINK TL-WR882N v1 support"
+	select SOC_QCA956X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_WMAC
+	select ATH79_DEV_USB
+
 endmenu
 
 config SOC_AR71XX
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -181,6 +181,7 @@
 	ATH79_MACH_TL_WR941ND_V5,	/* TP-LINK TL-WR941ND v5 */
 	ATH79_MACH_TL_WR941N_V6,	/* TP-LINK TL-WR941N v6 */
 	ATH79_MACH_TL_WR941N_V7,	/* TP-LINK TL-WR941N v7 */
+	ATH79_MACH_TL_WR882N_V1,	/* TP-LINK TL-WR882N v1 */
 	ATH79_MACH_TUBE2H,		/* Alfa Network Tube2H */
 	ATH79_MACH_UBNT_AIRGW,		/* Ubiquiti AirGateway */
 	ATH79_MACH_UBNT_AIRGWP,		/* Ubiquiti AirGateway Pro */
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -150,6 +150,7 @@
 obj-$(CONFIG_ATH79_MACH_TL_WR2543N)	+= mach-tl-wr2543n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR703N)	+= mach-tl-wr703n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR720N_V3)	+= mach-tl-wr720n-v3.o
+obj-$(CONFIG_ATH79_MACH_TL_WR882N_V1)	+= mach-tl-wr882n-v1.o
 obj-$(CONFIG_ATH79_MACH_TUBE2H)		+= mach-tube2h.o
 obj-$(CONFIG_ATH79_MACH_UBNT)		+= mach-ubnt.o
 obj-$(CONFIG_ATH79_MACH_UBNT_XM)	+= mach-ubnt-xm.o
